### PR TITLE
Add ST_Points spatial function

### PR DESCRIPTION
**Pull request by cooldoger**
Add ST_Points(geometry) function to return an array of points in a
LineString geometry.

### CODE REVIEW

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 1]
```diff
@@ -884,6 +884,26 @@ public static Slice stEndPoint(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+    @SqlType("array(" + GEOMETRY_TYPE_NAME + ")")
+    public static Block stPoints(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        validateType("ST_Points", geometry, EnumSet.of(LINE_STRING));
+        if (geometry.isEmpty()) {
+            return null;
+        }
+        MultiPath lines = (MultiPath) geometry.getEsriGeometry();
+        SpatialReference reference = geometry.getEsriSpatialReference();
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Spatial reference is always null, hence, drop this variable is use `null` below.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 2]
```diff
@@ -997,6 +1017,7 @@ public static Slice stExteriorRing(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
     @SqlType(GEOMETRY_TYPE_NAME)
     public static Slice stIntersection(@SqlType(GEOMETRY_TYPE_NAME) Slice left, @SqlType(GEOMETRY_TYPE_NAME) Slice right)
     {
+
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Unnecessary change.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 3]
```diff
@@ -519,6 +519,14 @@ public void testSTStartEndPoint()
         assertInvalidFunction("ST_AsText(ST_EndPoint(ST_GeometryFromText('POLYGON ((2 0, 2 1, 3 1))')))", "ST_EndPoint only applies to LINE_STRING. Input type is: POLYGON");
     }
 
+    @Test
+    public void testSTPoints()
+    {
+        assertFunction("transform(ST_Points(ST_GeometryFromText('LINESTRING (8 4)')), x -> ST_AsText(x))", new ArrayType(VARCHAR), ImmutableList.of("POINT (8 4)", "POINT (8 4)"));
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: All spatial functions assume that input geometries are well-formed. The functions don't include explicit checks because these are expensive. `LINESTRING (8 4)` is not a valid geometry, hence, let's not test it here.

Let's add a test for empty linestring and a closed linestring (a ring) though.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 4]
```diff
@@ -519,6 +519,14 @@ public void testSTStartEndPoint()
...
+        assertFunction("transform(ST_Points(ST_GeometryFromText('LINESTRING (8 4, 3 9, 5 6)')), x -> ST_AsText(x))", new ArrayType(VARCHAR), ImmutableList.of("POINT (8 4)", "POINT (3 9)", "POINT (5 6)"));
+        assertInvalidFunction("transform(ST_Points(ST_GeometryFromText('POLYGON ((8 4, 3 9, 5 6))')), x -> ST_AsText(x))", "ST_Points only applies to LINE_STRING. Input type is: POLYGON");
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: `transform` is unnecessary here

`assertInvalidFunction("ST_Points(ST_GeometryFromText('POLYGON ((8 4, 3 9, 5 6))'))", "ST_Points only applies to LINE_STRING. Input type is: POLYGON");`

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 5]
```diff
@@ -519,6 +519,14 @@ public void testSTStartEndPoint()
         assertInvalidFunction("ST_AsText(ST_EndPoint(ST_GeometryFromText('POLYGON ((2 0, 2 1, 3 1))')))", "ST_EndPoint only applies to LINE_STRING. Input type is: POLYGON");
     }
 
+    @Test
+    public void testSTPoints()
+    {
+        assertFunction("transform(ST_Points(ST_GeometryFromText('LINESTRING (8 4)')), x -> ST_AsText(x))", new ArrayType(VARCHAR), ImmutableList.of("POINT (8 4)", "POINT (8 4)"));
+        assertFunction("transform(ST_Points(ST_GeometryFromText('LINESTRING (8 4, 3 9, 5 6)')), x -> ST_AsText(x))", new ArrayType(VARCHAR), ImmutableList.of("POINT (8 4)", "POINT (3 9)", "POINT (5 6)"));
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: nit: consider extracting a helper method similar to `assertSTGeometries`:

`assertSTPoints("LINESTRING (8 4, 3 9, 5 6)", "8 4", "3 9", "5 6")`

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 6]
```diff
@@ -316,6 +316,10 @@ Accessors
     Returns the last point of a LineString geometry as a Point.
     This is a shortcut for ST_PointN(geometry, ST_NumPoints(geometry)).
 
+.. function:: ST_Points(Geometry) -> array(Point)
```
Comment (mbasmanova) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: In the docs functions are listed in alphabetical order. ST_Points goes right after ST_PointN.

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 7]
```diff
@@ -316,6 +316,10 @@ Accessors
...
+
+    Returns an array of points in a LineString geometry.
```
Comment (mbasmanova) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: s/LineString geometry/linestring

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java (hunk 2 above)
Comment (cooldoger) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Removed. This is causing check-style fail.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java (hunk 1 above)
Comment (mbasmanova) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: @cooldoger That's correct. If you'd like to remove spatial references from other places, do so in a separate commit.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 8]
```diff
@@ -519,6 +519,23 @@ public void testSTStartEndPoint()
         assertInvalidFunction("ST_AsText(ST_EndPoint(ST_GeometryFromText('POLYGON ((2 0, 2 1, 3 1))')))", "ST_EndPoint only applies to LINE_STRING. Input type is: POLYGON");
     }
 
+    @Test
+    public void testSTPoints()
+    {
+        assertSTPoints("LINESTRING EMPTY");
+        assertSTPoints("LINESTRING (0 0, 0 0)", "(0 0)", "(0 0)");
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: nit: perhaps, remove parentheses from points, e.g. `assertSTPoints("LINESTRING (0 0, 0 0)", "0 0", "0 0");`

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java (hunk 1 above)
Comment (cooldoger) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Here is a a separate PR for that: #12309

**Comment by mbasmanova** [comment]
@cooldoger Jay, thank you for the contribution.
